cpp/arcticdb/column_store/column.cpp

```diff
@@ -220,66 +220,65 @@ void Column::append(const Column& other, position_t at_row) {
                 row_count(), sparse_map().count());
 }
 
+void Column::physical_sort_external(std::vector<uint32_t> &&sorted_pos) {
+    size_t physical_rows = row_count();
+    auto& buffer = data_.buffer();
+
+    util::check(sorted_pos.size() == physical_rows, "Mismatch between sorted_pos size and row_count: {} != {}",
+        sorted_pos.size(), physical_rows);
+
+    type().visit_tag([&buffer, &sorted_pos, &physical_rows] (auto tdt) {
+        using TagType = decltype(tdt);
+        using RawType = typename TagType::DataTypeTag::raw_type;
+
+        for (auto i=0u; i<physical_rows; ++i){
+            if (i != sorted_pos[i]){
+                auto& current = buffer.cast<RawType>(i);
```

Comment:
You can eliminate the cost of the buffer.cast to i here by iterating over the physical blocks, since we need random access to the target but we know we're just forward-iterating the source. Alternatively we wait until we have a smarter iterator that would do this for us as described in the other comment.

Reply:
I think the cost of the buffer.cast to i is quite negligible and the added complexity to optimize it won't be worth it (and might make things slower).

The buffer.cast(i) will be called once for each permutation cycle (of length>1) in sorted_pos.

For random permutations the expected number of cycles is O(logn) see [here](https://en.wikipedia.org/wiki/Random_permutation_statistics#Expected_number_of_cycles_of_any_length_of_a_random_permutation).

So in the average case this would optimize O(logn) out of the O(n) calls to buffer.cast. In fact to optimize these O(logn) calls to cast we would need to do O(n) moves on the smarter iterator which might turn out to be slower.

Reply:
Right, I see. So since we often don't need the value of the current item at i, we'd end up having to calculate a random skip forward in the buffer anyway (i.e. the next time i != sorted_pos[i]), so we'd pretty much end up doing the same calculation buffer.cast does (or we'd end up doing a lot of branches to see if this was the end of the block, which as you say might be slower.

Cool, this looks good - nice work!